THREESCALE-7497: Removal of 3scale Istio adapter docs to add 3scale WebAssembly docs. (#36249)
openshift/openshift-docs: modules/ossm-configuring-the-threescale-wasm-auth-module.adoc

// Module included in the following assembly:
//
// service_mesh/v2x/ossm-threescale-webassembly-module.adoc

[id="ossm-configuring-the-threescale-wasm-auth-module_{context}"]
= Configuring the `threescale-wasm-auth` module

NOTE: To configure the 3scale `threescale-wasm-auth` module, refer to {ProductName} custom resources for instructions on adding adapter parameters to the custom resource file.

.The Service Mesh extension

{ProductShortName} provides a link:https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources[custom resource definition] to specify and apply _Proxy-WASM_ extensions to sidecar proxies. This custom resource is referred to as link:https://docs.openshift.com/container-platform/4.8/service_mesh/v2x/ossm-extensions.html[`ServiceMeshExtension`]. This custom resource will be applied to a set of workloads that require HTTP API management with 3scale.

.Prerequisites

* Identify a kubernetes workload and namespace on your {ProductShortName} deployment that you will apply this module.

> * You must have a 3scale account (link:https://www.3scale.net/signup[SaaS] or link:https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.11/html-single/installing_3scale/index#install-threescale-on-openshift-guide[3scale 2.11 On-Premises]) with a matching service and relevant applications and metrics defined.

Author: We refer to then as tenant account. I will amend here.

* If you apply the module to the `productpage` microservice in the `bookinfo` namespace, see the link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.8/html-single/service_mesh/index#ossm-tutorial-bookinfo-overview_ossm-create-mesh[Bookinfo sample application]. The following example is the format for the custom resource for `threescale-wasm-auth`:
+
----
apiVersion: maistra.io/v1alpha1
kind: ServiceMeshExtension
metadata:
  name: threescale-wasm-auth
  namespace: bookinfo
spec:
  workloadSelector:
    labels:
      app: productpage
  config: <yaml_configuration>

>   image: quay.io/3scale/threescale-wasm-auth:1.0

Reviewer: This image is definitely wrong as I don't believe we will use quay for released image

Author: @pehala Thanks for catching this.

@rahulanand16nov Let me know where I can get the correct image path when it's avialable. Cheers.

Reviewer: @dfennessy It should be `registry.redhat.io/openshift-service-mesh/3scale-auth-wasm-rhel8`

// Module included in the following assembly:
//
// service_mesh/v2x/ossm-threescale-webassembly-module.adoc

[id="ossm-configuring-the-threescale-wasm-auth-module_{context}"]
= Configuring the `threescale-wasm-auth` module

> As the OpenShift administrator, configure the `threescale-wasm-auth` module to authorize HTTP requests to 3scale API Management through an application binary interface (ABI). The ABI defines the interaction between host and the module, exposing the hosts services, and allows you to use the module to process proxy requests.
> 

Reviewer 2: ```suggestion
Cluster administrators on {product-title} can configure the `threescale-wasm-auth` module to authorize HTTP requests to 3scale API Management through an application binary interface (ABI). The ABI defines the interaction between host and the module, exposing the hosts services, and allows you to use the module to process proxy requests.

```
suggestion, sorry I missed this before; where possible use the product name / version attributes https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#product-name-and-version

> == The Service Mesh extension

Reviewer 2: needs an ID?

Author: I'm not sure. Doesn't an ID usually need a corresponding xref or include? And each new subheading is now automatically appearing in the TOC. Is this necessary or can I exclude them using the [discrete] tag?

Author: @abrennan89 I just have this outstanding couple of questions for the moment regarding IDs. It's a quick update and then I'll submit the PR to the squad for more eyes on if you're happy with the updates I've made to your reviews. 

Thank you for all your fantastic review comments.

Reviewer 2: > Doesn't an ID usually need a corresponding xref or include?

No, I believe it's for all headings and titles.
"All titles and section headings must have an anchor ID. The anchor ID must be similar to the title or section heading."
https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-titles-and-section-headings

> And each new subheading is now automatically appearing in the TOC. Is this necessary or can I exclude them using the [discrete] tag?

Yes, you should be able to use the [discrete] tag to hide them

Thanks for the updates! 🙂

Author: Thanks @abrennan89 :+1:

{ProductShortName} provides a link:https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources[custom resource definition] to specify and apply _Proxy-WASM_ extensions to sidecar proxies, known as link:https://docs.openshift.com/container-platform/4.8/service_mesh/v2x/ossm-extensions.html[`ServiceMeshExtension`]. {ProductShortName} applies this custom resource to the set of workloads that require HTTP API management with 3scale.

[NOTE]
====
Configuring the the WebAssembly extension is currently a manual process. Support for fetching the configuration for services from the 3scale system will be available in a future release.
====

.Prerequisites

> * Identify a kubernetes workload and namespace on your {ProductShortName} deployment that you will apply this module.

Reviewer 2: Should Kubernetes have a capital letter?

[NOTE]
====
Configuring the the WebAssembly extension is currently a manual process. Support for fetching the configuration for services from the 3scale system will be available in a future release.
====

.Prerequisites

* Identify a Kubernetes workload and namespace on your {ProductShortName} deployment that you will apply this module.
* You must have a 3scale tenant account. See link:https://www.3scale.net/signup[SaaS] or link:https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.11/html-single/installing_3scale/index#install-threescale-on-openshift-guide[3scale 2.11 On-Premises]) with a matching service and relevant applications and metrics defined.
* If you apply the module to the `productpage` microservice in the `bookinfo` namespace, see the link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.8/html-single/service_mesh/index#ossm-tutorial-bookinfo-overview_ossm-create-mesh[Bookinfo sample application].
** The following example is the YAML format for the custom resource for `threescale-wasm-auth`.

> This example refers to the Maistra, upstream version of {ProductShortName}, `ServiceMeshExtension` API. You must declare the namespace where the `threescale-wasm-auth` module is deployed, alongside a `WorkloadSelector` to identify the set of applications the module will apply to.

Author: Great idea. Thank you.

+
[source,yaml]
----
apiVersion: maistra.io/v1
kind: ServiceMeshExtension
metadata:
  name: threescale-wasm-auth
  namespace: bookinfo
spec:
  workloadSelector:
    labels:
      app: productpage
  config: <yaml_configuration>
  image: registry.redhat.io/openshift-service-mesh/3scale-auth-wasm-rhel8
  phase: PostAuthZ
  priority: 100
----
+
The `spec.config` field depends on the module configuration and it is not populated in the previous example. Instead, the example uses the <yaml_configuration> placeholder value. You can use the format of this custom resource example.
+
The `spec.config` field varies depending on the application. All other fields persist across multiple instances of this custom resource. As examples:
+
* `image`:  Only changes when newer versions of the module are deployed.
* `phase`:  Remains the same, since this module needs to be invoked after the proxy has done any local authorization, such as validating OpenID Connect (OIDC) tokens.

> Once you have the module configuration in `spec.config` and the rest of the custom resource, apply it with the `oc apply` command:

Author: Great! Thank you :)

// Module included in the following assembly:
//
// service_mesh/v2x/ossm-threescale-webassembly-module.adoc

[id="ossm-configuring-the-threescale-wasm-auth-module_{context}"]
= Configuring the threescale-wasm-auth module

Cluster administrators on {product-title} can configure the `threescale-wasm-auth` module to authorize HTTP requests to 3scale API Management through an application binary interface (ABI). The ABI defines the interaction between host and the module, exposing the hosts services, and allows you to use the module to process proxy requests.

[id="the-service-mesh-extension_{context}"]
== The Service Mesh extension

> {ProductShortName} provides a xref:../../operators/understanding/crds/crd-extending-api-with-crds.adoc#crd-extending-api-with-crds[custom resource definition] to specify and apply Proxy-WASM extensions to sidecar proxies, known as xref:../../service_mesh/v2x/ossm-extensions.adoc[`ServiceMeshExtension`]. {ProductShortName} applies this custom resource to the set of workloads that require HTTP API management with 3scale.

Reviewer 3: ```
xref:../../service_mesh/v2x/ossm-extensions.adoc[`ServiceMeshExtension`]
```

Try putting the anchor on the end here:

```
xref:../../service_mesh/v2x/ossm-extensions.adoc#ossm-extensions[`ServiceMeshExtension`]
```

[NOTE]
====
Configuring the the WebAssembly extension is currently a manual process. Support for fetching the configuration for services from the 3scale system will be available in a future release.
====

.Prerequisites

* Identify a Kubernetes workload and namespace on your {ProductShortName} deployment that you will apply this module.
* You must have a 3scale tenant account. See link:https://www.3scale.net/signup[SaaS] or link:https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.11/html-single/installing_3scale/index#install-threescale-on-openshift-guide[3scale 2.11 On-Premises] with a matching service and relevant applications and metrics defined.

> * If you apply the module to the `productpage` microservice in the `bookinfo` namespace, see the xref:ossm-tutorial-bookinfo-overview_ossm-create-mesh[Bookinfo sample application].

Reviewer 3: Looks like this might need a full path? Is it this assembly:

```
xref:../../service_mesh/v2x/prepare-to-deploy-applications-ossm.adoc#ossm-tutorial-bookinfo-overview_ossm-create-mesh[Bookinfo sample application]
```

** The following example is the YAML format for the custom resource for `threescale-wasm-auth` module.
This example refers to the upstream Maistra version of {ProductShortName}, ServiceMeshExtension API. You must declare the namespace where the `threescale-wasm-auth` module is deployed, alongside a `WorkloadSelector` to identify the set of applications the module will apply to:
+
[source,yaml]
----
apiVersion: maistra.io/v1
kind: ServiceMeshExtension
metadata:
  name: threescale-wasm-auth
  namespace: bookinfo <1>
spec:
  workloadSelector: <2>
    labels:
      app: productpage
  config: <yaml_configuration>

>   image: registry.redhat.io/openshift-service-mesh/3scale-auth-wasm-rhel8

Reviewer: The released image is `registry.redhat.io/openshift-service-mesh/3scale-auth-wasm-rhel8:0.0.1`
